### DISCUSSION TITLE
Hi, excuse me. Can you answer a few questions for me

### DISCUSSION THREAD

**Discussion by guminlianren** [bcakmakoglu/vue-flow#1572] [Q&A]
1、VueFlow has added apply-default, causing the links between lines to become ineffective. I have configured onConnect(addEdges), and it has been mentioned that this property will be removed in the next version without indicating a replacement.

2、Is there an API that does not have zoom in/out functionality and background movement? Thank you

**Comment by guminlianren** [comment]
3、 After connecting, the `onConnect` function is called with `(connection) => {
    addEdges(connection);
  });`. How can I dynamically add `markerEnd` to this?  Or create arrows dynamically

**Comment by bcakmakoglu** [comment]
> VueFlow has added apply-default, causing the links between lines to become ineffective.

`apply-default` doesn't make any "link" (you mean edges?) "ineffective" (what is this supposed to mean?).

> this property will be removed in the next version without indicating a replacement.

There won't be a replacement for it. It's going to be removed and that's it, no more "auto" application of changes.

> Is there an API that does not have zoom in/out functionality and background movement?

No idea what this is supposed to mean.

> How can I dynamically add markerEnd

Have you tried quite literally just adding a markerEnd to the edge? ^^

```ts
const onConnect = (conn) => {
	addEdges({ ...conn, markerEnd: MarkerType.Arrow })
}
```

**Comment by guminlianren** [comment]
Thank you for your reply. Here's the translation:

1、I have a requirement where I need to ask for confirmation before deleting. When the user deletes using the delete key, there is no confirmation. Can this confirmation functionality be added? After adding this property, the connecting lines in the background automatically disappear.

2、Are there properties to disable zooming, scaling, and moving the background?

3、If dynamically created components are maintained within a component, how can I access them from outside?

Let me know if you need further assistance with these topics!

**Reply by guminlianren** [reply]
I know that by configuring :min-zoom="1", :max-zoom="1", and :default-viewport="{ zoom: 1 }", the zoom functionality will be disabled. How can I also prevent the canvas from being moved?

**Reply by bcakmakoglu** [reply]
1. [There's an example for this.](https://vueflow.dev/examples/confirm.html). `applyDefault` will be removed, yes, but the concept of applying changes will remain, it's just not going to be auto-applied under the hood.
2. [Read the docs on viewport related props.](https://vueflow.dev/guide/vue-flow/config.html#viewport-options)
3. No idea what that means.

**Reply by guminlianren** [reply]
Thank you for your reply. Here's the translation:" 1、Yes, it's exactly as per your example. Are there any other ways to prevent the keyboard 'del' event from deleting nodes?"

2、"I dynamically create a line like this using 'onConnect((connection) => { addEdges({ ...connection, markerEnd: MarkerType.Arrow }); });'. Later, when I use the 'nextTick' function to print 'edges', I find it empty. If I need to obtain 'edges' through a function, help me translate this into English."

Let me know if there's anything else you need help with!

**Reply by bcakmakoglu** [reply]
1. yep, set `:delete-key-code="null"` which will disable the delete key code entirely.
2. What? 🤔

> Let me know if there's anything else you need help with!

Very nice of you to say but I think your translation might be flipped since you're the one asking for help not me 😂

**Reply by guminlianren** [reply]
"I create a line dynamically using 'onConnect((connection) => { addEdges({ ...connection, markerEnd: MarkerType.Arrow }); });'. Subsequently, when I use the 'nextTick' function to print 'edges', I notice it's empty. If I need to retrieve 'edges' using a function,“，I copied it. I'm sorry

**Reply by bcakmakoglu** [reply]
You can use this to get edges:
```ts
const { edges } = useVueFlow()

edges.value
```

If you're not using `v-model:edges="edges"` changes won't be propagated to your edges array, so you'll have to get them from the internal state instead.

**Reply by guminlianren** [reply]
You configured it like this: :min-zoom="1", :max-zoom="2", :default-viewport="{ zoom: 1 }". Why does it still zoom to twice the size after refreshing?

**Reply by bcakmakoglu** [reply]
I assume you're using fit view on init, which will, as the name implies, fit the view.

**Reply by guminlianren** [reply]
I'm not quite sure what you mean

**Reply by bcakmakoglu** [reply]
You're using `fit-view-on-init` and as I said, this prop will *fit the view* on initialization. 
That means that VueFlow will try to fit around *all nodes* so that they're in the visible frame, which can result in the zoom level being changed.

**Comment by guminlianren** [comment]
https://vueflow.dev/examples/layout.html Hello, this example is too complicated, I don't need the animation in it. I need to automatically layout my nodes, there is no concise version of the code, thank you。
![1723171286946](https://github.com/user-attachments/assets/8c744125-4ae0-48bc-8bb6-d95ec2f2a2ce)

**Reply by guminlianren** [reply]
onConnect((connection) => {
    addEdges({ ...connection, markerEnd: MarkerType.Arrow, type: 'animation' });
  });I configured it this way. Why isn't it a straight line? Interface: DefaultEdgeTypes How do I use this?

**Reply by bcakmakoglu** [reply]
>  I need to automatically layout my nodes, there is no concise version of the code, thank you。 

Read the example properly, please.
Everything you need for layouting is in `useLayout`.

I'm not writing an extra example just for you, sorry.

> Why isn't it a straight line? 

Why would it be? You haven't used `type: 'straight'`.

**Comment by guminlianren** [comment]
Ok, I'll look at the examples carefully  。https://vueflow.dev/examples/layout.html   Because that's what we're using in this example  "  animated    "{ id: 'e1-2', source: '1', target: '2', type: edgeType, animated: true },  I thought animated was a straight line  ,
![image](https://github.com/user-attachments/assets/a475bbde-38a4-417f-acd1-605114bef27b)  How do you do this straight line    ?
![image](https://github.com/user-attachments/assets/aaea3cbd-7c49-480b-a494-3db92443093d)
Why is my place like this

**Reply by bcakmakoglu** [reply]
Why would `animated` mean straight line? 😅 It means "the line is animated".
If you want a straight line you simply use `type: 'straight'` as I said previously.

>  Why is my place like this

I don't know, you haven't given me any code or reproduction of your code to inspect.

**Comment by guminlianren** [comment]
1、  
![image](https://github.com/user-attachments/assets/1483da6b-3979-4bb0-9dec-914d684fefc8)
  I want to add nodes up and down, left and right, so I'll write it like this. Click on the right line from above?
![image](https://github.com/user-attachments/assets/62beb7cb-7596-43aa-82a8-e28d038ee5e4)   2、 set :delete-key-code="null"        <VueFlow
      ref="VueFlowRef"
      :nodes="nodes"
      :edges="edges"
      :delete-key-code="null"
      @keydown="handleKeydown"
    >
Since then, when I monitored the keyboard incident, edges can't. Why? nodes can be monitored, so I can't delete this line

**Reply by bcakmakoglu** [reply]
Read the docs: https://vueflow.dev/guide/handle.html#multiple-handles

**Comment by guminlianren** [comment]
node-Toolbar can't be used by edges, right? What if edges want a toolbar

**Reply by bcakmakoglu** [reply]
Node toolbar is only for nodes *but* you can use the `<EdgeLabelRenderer>` to create a toolbar-like menu.

[Check this example](https://vueflow.dev/examples/edges/) and open up the `EdgeWithButton.vue` file :)

**Comment by guminlianren** [comment]
To use slots, you must have type: custom. I want to keep type: 'step'. In this case, how to use slots? All lines can use slots

**Reply by bcakmakoglu** [reply]
`custom` is just an example key-word. Your type can be anything, you could call your custom edge type `deezNuts` and it's still ok.

If you want to use a custom edge as a (smooth)step edge, you can implement it using `getSmoothStepPath` which is exported by `@vue-flow/core` (or just use the `<SmoothStepEdge>` component that is also exported.

```vue
<VueFlow ...>
	<template #edge-deezNuts="props">
		<DeezNutsEdge v-bind="props />
	</template>
</VueFlow>
```

```vue
// DeezNutsEdge.vue
<script lang="ts" setup>
import type { EdgeProps } from '@vue-flow/core'
import { getSmoothStepPath } from '@vue-flow/core'

const props = defineProps<EdgeProps>()

const path = computed(() => getSmoothStepPath({
	...props,
	borderRadius: 0 // use 0 as the borderRadius if you want to have a *step* edge instead of a smoothstep one
}))
</script>

<script lang="ts">
export default {
  inheritAttrs: false,
}
</script>

<template>
  <path :id="id" :style="style" class="vue-flow__edge-path" :d="path[0]" :marker-end="markerEnd" />
</template>
```

**Comment by guminlianren** [comment]
I don't want the node to be out of view of the current viewer? Do not want the background to drag (move)

**Reply by bcakmakoglu** [reply]
Please *read the docs*: https://vueflow.dev/guide/vue-flow/config.html#viewport-options

I am not your personal documentation chat-bot.
99% of what you're asking is documented, please *READ* the documentation first!

**Comment by bcakmakoglu** [comment]
I am closing this discussion since it devolved into just me answering questions that basically are all documented already.
If you have a specific issue that you can't handle even after having read the docs, open a new topic with a specific question.
